http-common/src/connector.rs

```diff
@@ -72,36 +68,174 @@ impl Incoming {
                 user_state,
             } => loop {
                 let (unix_stream, _) = listener.accept().await?;
-                let ucred = unix_stream.peer_cred()?;
-                let user_state = user_state
-                    .entry(ucred.uid())
-                    .or_insert_with(|| {
-                        std::sync::Arc::new(tokio::sync::Semaphore::new(MAX_REQUESTS_PER_USER))
-                    })
-                    .clone();
+                let (user_state, ucred) = get_user_auth(&unix_stream, user_state)?;
 
                 let server = crate::uid::UidService::new(ucred.pid(), ucred.uid(), server.clone());
                 tokio::spawn(async move {
-                    match user_state.try_acquire_owned() {
-                        Ok(_permit) => {
-                            if let Err(http_err) = hyper::server::conn::Http::new()
-                                .serve_connection(unix_stream, server)
-                                .await
-                            {
-                                log::info!("Error while serving HTTP connection: {}", http_err);
-                            }
-                        }
-                        Err(limit_err) => {
-                            log::info!(
-                                "Error while acquiring permit for HTTP connection: {}",
-                                limit_err
-                            );
-                        }
-                    }
+                    serve_unix(user_state, server, unix_stream).await;
                 });
             },
         }
     }
+
+    #[cfg(feature = "serve_with_shutdown")]
+    pub async fn serve_with_shutdown<H>(
+        &mut self,
+        server: H,
+        shutdown: tokio::sync::oneshot::Receiver<()>,
+    ) -> std::io::Result<()>
+    where
```

Comment:
@arsing Is there a way to collapse these `where` clauses so I don't have to attach it to every function that takes H? Just curious, but I didn't find any.

Reply:
Until we get trait aliases (`trait Foo = Bar + Baz`, RFC opened in 2017 so it'll probably get stabilized by 3017 AD), the only way is to make a new trait that inherits the bounds and provide a blanket impl for existing types: `trait Foo: Bar + Baz { } impl<T> Foo for T where T: Bar + Baz { }` and then use `: Foo` as the bound.